File: 797.py
class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        n = len(graph)
        target = n - 1
        result = []
        stack = [(0,[0])]

        while stack:
            node, path = stack.pop()
            if node == target:
                result.append(path)
                continue
            for neighbor in reversed(graph[node]):
                stack.append((neighbor,path + [neighbor]))

        return result

"""
graph = [[4,3,1],[3,2,4],[3],[4],[]]
solution = Solution()
result = solution.allPathsSourceTarget(graph)
print(result)
"""

File: 797.cpp
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> res;
    std::vector<int> way;
    int n;
    void dfs(std::vector<std::vector<int>>& graph, int v) {
        for (const int& p : graph[v]) {
            way.push_back(p);
            if (p == n - 1) {
                res.push_back(way);
            }
            else {
                dfs(graph, p);
            }
            way.pop_back();
        }
    }
    std::vector<std::vector<int>> allPathsSourceTarget(std::vector<std::vector<int>>& graph) {
        n = graph.size();
        way.push_back(0);
        dfs(graph, 0);
        return res;
    }
};
